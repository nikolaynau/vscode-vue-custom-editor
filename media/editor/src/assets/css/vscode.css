@import "vscode-theme";
@import "vscode-default";
