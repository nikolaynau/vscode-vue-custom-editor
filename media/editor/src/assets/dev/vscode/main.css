@import 'vscode-default/all';

body {
  background-color: var(--vscode-editor-background);
}
