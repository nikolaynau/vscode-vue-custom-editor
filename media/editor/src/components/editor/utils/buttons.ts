export interface ButtonDefinition {
  id: number;
  label: string;
  value: number;
  side: 'left' | 'right';
}

export function createButtons(): Array<ButtonDefinition> {
  return [
    {
      id: 1,
      label: '-10',
      value: -10,
      side: 'left'
    },
    {
      id: 2,
      label: '-5',
      value: -5,
      side: 'left'
    },
    {
      id: 3,
      label: '-1',
      value: -1,
      side: 'left'
    },
    {
      id: 4,
      label: '+1',
      value: 1,
      side: 'right'
    },
    {
      id: 5,
      label: '+5',
      value: 5,
      side: 'right'
    },
    {
      id: 6,
      label: '+10',
      value: 10,
      side: 'right'
    }
  ];
}
