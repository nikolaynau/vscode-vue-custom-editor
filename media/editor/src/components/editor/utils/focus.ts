export function nextFocusElement(): void {
  const activeElement = document.activeElement;
  if (
    !activeElement?.classList.contains('v-editor__input') &&
    !activeElement?.classList.contains('v-action') &&
    !activeElement?.classList.contains('v-button')
  ) {
    return;
  }
  let nextElement: Element | null | undefined =
    activeElement?.nextElementSibling;
  if (nextElement?.classList.contains('v-editor__controls')) {
    nextElement = nextElement.children[0];
  } else if (
    !nextElement &&
    activeElement?.parentElement?.nextElementSibling?.classList.contains(
      'v-editor__input'
    )
  ) {
    nextElement = activeElement?.parentElement?.nextElementSibling;
  }
  if (!nextElement) {
    nextElement =
      activeElement?.parentNode?.parentNode?.children[0]?.children[0];
  }

  if (nextElement) {
    (nextElement as HTMLElement).focus();
  }
}

export function prevFocusElement(): void {
  const activeElement = document.activeElement;
  if (
    !activeElement?.classList.contains('v-editor__input') &&
    !activeElement?.classList.contains('v-action') &&
    !activeElement?.classList.contains('v-button')
  ) {
    return;
  }
  let prevElement: Element | null | undefined =
    activeElement?.previousElementSibling;
  if (prevElement?.classList.contains('v-editor__controls')) {
    prevElement = prevElement.lastElementChild;
  } else if (
    !prevElement &&
    activeElement?.parentElement?.previousElementSibling?.classList.contains(
      'v-editor__input'
    )
  ) {
    prevElement = activeElement?.parentElement?.previousElementSibling;
  }
  if (!prevElement) {
    prevElement =
      activeElement?.parentNode?.parentNode?.lastElementChild?.lastElementChild;
  }

  if (prevElement) {
    (prevElement as HTMLElement).focus();
  }
}
