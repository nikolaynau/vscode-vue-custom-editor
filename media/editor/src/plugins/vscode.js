export default {
  install(app) {
    const vscode = window?.acquireVsCodeApi?.();
    app.config.globalProperties.$vscode = vscode;
    app.provide("$vscode", vscode);
  }
}
