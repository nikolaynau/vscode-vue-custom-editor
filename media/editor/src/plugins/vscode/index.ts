import type { Plugin, App } from 'vue';
import { vscodeKey } from './injection-symbols';
import vscodeApi from './vscode-api';

export default {
  install(app: App) {
    const vscode =
      (import.meta.env.MODE === 'development' ||
        import.meta.env.MODE === 'test') &&
      typeof window.acquireVsCodeApi !== 'function'
        ? vscodeApi
        : window.acquireVsCodeApi();

    app.config.globalProperties.$vscode = vscode;
    app.provide(vscodeKey, vscode);
  }
} as Plugin;
