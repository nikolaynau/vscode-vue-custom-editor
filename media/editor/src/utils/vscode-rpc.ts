import logger from '@/utils/logger';
import { RpcProvider } from 'worker-rpc';

export class VsCodeRpc {
  public provider: RpcProvider;

  public constructor(vscode: VsCodeApi) {
    this.provider = new RpcProvider(message => vscode.postMessage(message));

    this.handleMessage = this.handleMessage.bind(this);
    this.handleError = this.handleError.bind(this);

    this.provider.error.addHandler(this.handleError);
    window.addEventListener('message', this.handleMessage);
  }

  private handleMessage(e: MessageEvent) {
    this.provider.dispatch(e.data);
  }

  private handleError(err: Error) {
    logger.error('[Rpc]: rpc provider error', err);
  }

  public destroy() {
    window.removeEventListener('message', this.handleMessage);
    const provider = this.provider as any;
    provider._dispatch = null;
    provider._rpcHandlers = [];
    provider._signalHandlers = {};
  }
}
