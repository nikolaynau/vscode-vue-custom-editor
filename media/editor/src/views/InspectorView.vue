<script setup lang="ts">
import { ref, onBeforeUnmount } from 'vue';
import type { InspectorDataModel } from '@/components/inspector/composables/use-inspector';
import { useVsCode } from '@/plugins/vscode/use-vscode';
import { VsCodeRpc } from '@/utils/vscode-rpc';
import type { EditCommandArray } from '@/components/editor/utils/types';
import VInspector from '@/components/inspector/VInspector.vue';

const dataModel = ref<InspectorDataModel | undefined>(undefined);
const vscode = useVsCode();

const rpc = new VsCodeRpc(vscode);
rpc.provider.registerRpcHandler('setData', setData);

function setData(data: InspectorDataModel) {
  dataModel.value = data;
}

function ready() {
  rpc.provider.signal('ready');
}

function onEdit(edits: EditCommandArray) {
  rpc.provider.signal('edit', edits);
}

onBeforeUnmount(() => {
  rpc.destroy();
});

ready();
</script>

<template>
  <div class="inspector-container">
    <VInspector :data-model="dataModel" @edit="onEdit" />
  </div>
</template>

<style>
@import 'inspector-view';
</style>
