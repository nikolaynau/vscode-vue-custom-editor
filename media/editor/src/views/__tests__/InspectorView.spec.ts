import {
  describe,
  it,
  expect,
  vi,
  afterEach,
  beforeEach,
  type Mock
} from 'vitest';
import { mount } from '@vue/test-utils';
import { nextTick } from 'vue';
import InspectorView from '../InspectorView.vue';
import { VsCodeRpc } from '@/utils/vscode-rpc';
import type { InspectorDataModel } from '@/components/inspector/composables/use-inspector';
import type {
  ChangeButtonValueCommand,
  EditCommandArray,
  ReplaceValueCommand
} from '@/components/editor/utils/types';
import vscode from '@/plugins/vscode';

vi.mock('../../utils/vscode-rpc.ts', () => {
  const VsCodeRpc = vi.fn();
  VsCodeRpc.prototype.provider = {
    signal: vi.fn(),
    registerRpcHandler: vi.fn()
  };
  VsCodeRpc.prototype.destroy = vi.fn();
  return { VsCodeRpc };
});

describe('InspectorView', () => {
  beforeEach(() => {
    vi.useFakeTimers();
  });

  afterEach(() => {
    vi.restoreAllMocks();
  });

  it('send ready signal', () => {
    const wrapper = mount(InspectorView, { global: { plugins: [vscode] } });
    expect(wrapper.html()).toMatchSnapshot();

    expect((VsCodeRpc as any).mock.instances).toHaveLength(1);
    expect(
      (VsCodeRpc as any).mock.instances[0].provider.signal
    ).toBeCalledTimes(1);
    expect((VsCodeRpc as any).mock.instances[0].provider.signal).toBeCalledWith(
      'ready'
    );
  });

  it('setData rpc', async () => {
    const inspectorDataModel = {
      counterValue: 10,
      buttons: [
        { id: 1, value: 2 },
        { id: 2, value: 3 }
      ]
    } as InspectorDataModel;

    const rpcHandlers: Record<string, Function> = {};
    (
      VsCodeRpc.prototype.provider.registerRpcHandler as Mock
    ).mockImplementation((id, handler) => {
      rpcHandlers[id] = handler;
    });

    const wrapper = mount(InspectorView, { global: { plugins: [vscode] } });
    expect(typeof rpcHandlers['setData']).toBe('function');

    rpcHandlers['setData'](inspectorDataModel);
    await nextTick();

    expect(wrapper.html()).toMatchSnapshot();

    const inputs = wrapper.findAll('input');
    expect(inputs).toHaveLength(3);
    expect(+inputs[0].element.value).toBe(inspectorDataModel.counterValue);
    expect(+inputs[1].element.value).toBe(inspectorDataModel.buttons[0].value);
    expect(+inputs[2].element.value).toBe(inspectorDataModel.buttons[1].value);
  });

  it('edit signal', async () => {
    const inspectorDataModel = {
      counterValue: 10,
      buttons: [
        { id: 1, value: 2 },
        { id: 2, value: 3 }
      ]
    } as InspectorDataModel;

    const rpcHandlers: Record<string, Function> = {};
    (
      VsCodeRpc.prototype.provider.registerRpcHandler as Mock
    ).mockImplementation((id, handler) => {
      rpcHandlers[id] = handler;
    });

    const wrapper = mount(InspectorView, { global: { plugins: [vscode] } });
    expect(typeof rpcHandlers['setData']).toBe('function');

    rpcHandlers['setData'](inspectorDataModel);
    await nextTick();

    const inputs = wrapper.findAll('input');
    expect(inputs).toHaveLength(3);

    await inputs[0].setValue(10);
    await inputs[1].setValue(11);
    await inputs[2].setValue(12);

    vi.runAllTimers();

    await nextTick();

    expect((VsCodeRpc as any).mock.instances).toHaveLength(1);
    const signalSpy = (VsCodeRpc as any).mock.instances[0].provider
      .signal as Mock;
    expect(signalSpy).toBeCalledTimes(4);
    expect(signalSpy.mock.calls[0][0]).toBe('ready');

    expect(signalSpy.mock.calls[1][0]).toBe('edit');
    expect(signalSpy.mock.calls[1][1]).toEqual([
      { name: 'replace', payload: { value: 10 } } as ReplaceValueCommand
    ] as EditCommandArray);

    expect(signalSpy.mock.calls[2][0]).toBe('edit');
    expect(signalSpy.mock.calls[2][1]).toEqual([
      {
        name: 'change-button',
        payload: { btnId: 1, value: 11 }
      } as ChangeButtonValueCommand
    ] as EditCommandArray);

    expect(signalSpy.mock.calls[3][0]).toBe('edit');
    expect(signalSpy.mock.calls[3][1]).toEqual([
      {
        name: 'change-button',
        payload: { btnId: 2, value: 12 }
      } as ChangeButtonValueCommand
    ] as EditCommandArray);
  });

  it('call destroy rpc on unmount component', async () => {
    const wrapper = mount(InspectorView, { global: { plugins: [vscode] } });
    expect((VsCodeRpc as any).mock.instances).toHaveLength(1);
    expect((VsCodeRpc as any).mock.instances[0].destroy).toBeCalledTimes(0);

    wrapper.unmount();

    expect((VsCodeRpc as any).mock.instances[0].destroy).toBeCalledTimes(1);
  });
});
