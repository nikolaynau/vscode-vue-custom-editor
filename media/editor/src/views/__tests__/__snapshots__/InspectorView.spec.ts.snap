// Vitest Snapshot v1

exports[`InspectorView > send ready signal 1`] = `
"<div class=\\"inspector-container\\">
  <div class=\\"v-inspector\\">
    <div class=\\"v-fieldset v-fieldset--b-border v-fieldset--padding\\">
      <div class=\\"v-fieldset__label\\">General</div>
      <div class=\\"v-field-layout v-field-layout--vertical\\">
        <div class=\\"v-field-layout__label\\"><label class=\\"v-label v-label--truncate\\">Counter</label></div>
        <div class=\\"v-field-layout__value\\">
          <div class=\\"v-input-layout\\"><input class=\\"v-input v-input--icon\\" placeholder=\\"Enter number\\" type=\\"text\\" tabindex=\\"0\\"><span class=\\"v-input__icon\\"><span class=\\"codicon codicon-info\\"></span></span></div>
        </div>
      </div>
    </div>
    <div class=\\"v-fieldset v-fieldset--padding\\">
      <div class=\\"v-fieldset__label\\">Controls</div>
    </div>
  </div>
</div>"
`;

exports[`InspectorView > setData rpc 1`] = `
"<div class=\\"inspector-container\\">
  <div class=\\"v-inspector\\">
    <div class=\\"v-fieldset v-fieldset--b-border v-fieldset--padding\\">
      <div class=\\"v-fieldset__label\\">General</div>
      <div class=\\"v-field-layout v-field-layout--vertical\\">
        <div class=\\"v-field-layout__label\\"><label class=\\"v-label v-label--truncate\\">Counter</label></div>
        <div class=\\"v-field-layout__value\\">
          <div class=\\"v-input-layout\\"><input class=\\"v-input v-input--icon\\" placeholder=\\"Enter number\\" type=\\"text\\" tabindex=\\"0\\"><span class=\\"v-input__icon\\"><span class=\\"codicon codicon-info\\"></span></span></div>
        </div>
      </div>
    </div>
    <div class=\\"v-fieldset v-fieldset--padding\\">
      <div class=\\"v-fieldset__label\\">Controls</div>
      <div class=\\"v-field-layout v-field-layout--vertical\\">
        <div class=\\"v-field-layout__label\\"><label class=\\"v-label v-label--truncate\\">Button #1</label></div>
        <div class=\\"v-field-layout__value\\">
          <div class=\\"v-input-layout\\"><input class=\\"v-input\\" placeholder=\\"Enter number\\" type=\\"text\\" tabindex=\\"0\\">
            <!--v-if-->
          </div>
        </div>
      </div>
      <div class=\\"v-field-layout v-field-layout--vertical\\">
        <div class=\\"v-field-layout__label\\"><label class=\\"v-label v-label--truncate\\">Button #2</label></div>
        <div class=\\"v-field-layout__value\\">
          <div class=\\"v-input-layout\\"><input class=\\"v-input\\" placeholder=\\"Enter number\\" type=\\"text\\" tabindex=\\"0\\">
            <!--v-if-->
          </div>
        </div>
      </div>
    </div>
  </div>
</div>"
`;
